### DISCUSSION TITLE
What does viewportCoordsToSceneCoords method do?

### DISCUSSION THREAD

**Discussion by Marwa-Da** [excalidraw/excalidraw#6017] [Q&A] [answered]
Hi community,
I'm facing a difficulty understanding what viewportCoordsToSceneCoords method does! I tried to opt it out and compare the results before and after applying it but didn't get its goal.

**Comment by dwelle** [comment, marked as answer]
Hi!

`viewportCoordsToSceneCoords(coords, appState)` convert viewport-relative coordinates (the `clientX`/`clientY` you get from mouse/pointer events) to coordinates relative to the `[0,0]` coordinate of the excalidraw canvas (which by default is at the top left of the canvas).

It takes into account the canvas scroll position. So if you scroll down by 100px without moving your cursor, the cursor's scene coords will be `[0, 100]`.

https://codesandbox.io/s/confident-david-9vn3te?file=/src/App.js

Note: in the above example I'm also showing scroll position. At present the scroll offsets are "inverted". If you scroll down by 100px, the `appState.scrollY` will be `-100`, although the standard should be `100`. See https://github.com/excalidraw/excalidraw/issues/5965

EDIT: there seems to be a bug with scene coords calculation under specific conditions that was uncovered as part of creating the example above. See https://github.com/excalidraw/excalidraw/issues/6026

**Reply by Marwa-Da** [reply]
Thank you so much for your detailed reply and for the example. Now I got what's going on.
